File: 2026-qualifier/simulate.ts
import * as fs from "node:fs";
import * as yaml from "js-yaml";
import { Worker } from "worker_threads";
import * as os from "os";
import * as path from "path";

// --- Option Parsing ---
const args = process.argv.slice(2);
const options = {
  configFile: args.find(a => !a.startsWith('--')) || 'groupH.json',
  algo: args.find(a => a.startsWith('--algo='))?.split('=')[1] || 'scan',
  worker: args.find(a => a.startsWith('--worker='))?.split('=')[1] || 'threads',
  prng: args.find(a => a.startsWith('--prng='))?.split('=')[1] || 'xorshift32',
};

console.log('Running with options:', options);


interface TeamResult {
  direct: number;
  playoff: number;
  fail: number;
  overall: number;
}

interface TallyCount {
  direct: number;
  playoff: number;
  fail: number;
}

type ResultTable = Record<Team, TeamResult>;

interface Config {
  numberOfSimulations: number;
  teams: string[];
  elo: Record<string, number>;
  homeBonus: number;
  currentPoints: Record<string, number>;
  fixtures: [string, string][];
  drawR: number;
  playoffWinProb: number;
  teamIdx: Map<string, number>;
  basePts: number[];
  precomputedFixtures: [number, number, number, number, number][];
  eloWinProb(rA: number, rB: number): number;
  drawProb(deltaElo: number): number;
}

const configContent = fs.readFileSync(options.configFile, "utf-8");

let cfg: Config;
const ext = options.configFile.split(".").pop();

if (ext === "json") {
  cfg = JSON.parse(configContent);
} else if (ext === "yaml" || ext === "yml") {
  cfg = yaml.load(configContent) as Config;
} else {
  throw new Error(`Unsupported config file format: ${ext}`);
}

cfg.eloWinProb = (rA: number, rB: number) => 1 / (1 + Math.pow(10, (rB - rA) / 400));
cfg.drawProb = function (deltaElo: number) {
  const w = 1 / (1 + Math.pow(10, -deltaElo / 400));
  return 2 * w * (1 - w) * this.drawR;
};

cfg.teamIdx = new Map<string, number>();
cfg.basePts = [];
cfg.teams.forEach((team, i) => {
  cfg.teamIdx.set(team, i);
  cfg.basePts[i] = cfg.currentPoints[team];
});

cfg.precomputedFixtures = cfg.fixtures.map(([home, away]) => {
  const homeIdx = cfg.teamIdx.get(home)!;
  const awayIdx = cfg.teamIdx.get(away)!;
  const delta = cfg.elo[home] + cfg.homeBonus - cfg.elo[away];
  const pDraw = cfg.drawProb(delta);
  const pHome = (1 - pDraw) * cfg.eloWinProb(cfg.elo[home] + cfg.homeBonus, cfg.elo[away]);
  const pAway = 1 - pHome - pDraw;
  return [homeIdx, awayIdx, pHome, pDraw, pAway];
});

const TEAMS = cfg.teams;
type Team = (typeof TEAMS)[number];

async function simulate(): Promise<ResultTable> {
  const numCores = os.cpus().length;
  const simsPerWorker = Math.floor(cfg.numberOfSimulations / numCores);
  const promises: Promise<Record<Team, TallyCount>>[] = [];

  console.log(`Using ${numCores} cores for simulation...`);

  const serializableCfg = {
    ...cfg,
    teamIdx: Array.from(cfg.teamIdx.entries()),
  };
  delete (serializableCfg as any).eloWinProb;
  delete (serializableCfg as any).drawProb;

  for (let i = 0; i < numCores; i++) {
    const startIdx = i * simsPerWorker;
    const endIdx = i === numCores - 1 ? cfg.numberOfSimulations : startIdx + simsPerWorker;

    const workerPromise = new Promise<Record<Team, TallyCount>>((resolve, reject) => {
      const workerData = { cfg: serializableCfg, startIdx, endIdx, opts: options, workerId: i };

      if (options.worker === 'web') {
        const worker = new (globalThis as any).Worker(new URL('./simulate_web_worker.js', import.meta.url).href, { type: 'module' });
        worker.onmessage = (event: any) => {
          resolve(event.data);
          worker.terminate();
        };
        worker.onerror = (err: any) => reject(err);
        worker.postMessage(workerData);
      } else { // Default to threads
        const worker = new Worker(path.join(__dirname, 'simulate_worker.js'), { workerData });
        worker.on("message", resolve);
        worker.on("error", reject);
        worker.on("exit", (code) => {
          if (code !== 0) reject(new Error(`Worker stopped with exit code ${code}`));
        });
      }
    });
    promises.push(workerPromise);
  }

  const workerResults = await Promise.all(promises);

  const finalTally: Record<Team, TallyCount> = TEAMS.reduce(
    (acc, t) => ({ ...acc, [t]: { direct: 0, playoff: 0, fail: 0 } }),
    {} as Record<Team, TallyCount>
  );

  for (const workerResult of workerResults) {
    for (const t of TEAMS) {
      finalTally[t].direct += workerResult[t].direct;
      finalTally[t].playoff += workerResult[t].playoff;
      finalTally[t].fail += workerResult[t].fail;
    }
  }

  const res = {} as ResultTable;
  for (const t of TEAMS) {
    const d = finalTally[t].direct / cfg.numberOfSimulations;
    const p = finalTally[t].playoff / cfg.numberOfSimulations;
    const f = finalTally[t].fail / cfg.numberOfSimulations;
    res[t] = { direct: d, playoff: p, fail: f, overall: d + p * cfg.playoffWinProb };
  }

  return res;
}

const pct = (x: number) => (x * 100).toFixed(1) + "%";

function showTeamOdds(table: ResultTable): void {
  console.log(`\nQualification probabilities (${cfg.numberOfSimulations} sims):\n`);
  console.table(
    TEAMS.map(t => ({
      Team: t,
      Direct: pct(table[t].direct),
      Playoff: pct(table[t].playoff),
      Eliminated: pct(table[t].fail),
      Overall: pct(table[t].overall),
    }))
  );
}

function showMatchOdds(): void {
  console.log("\nOdds for each remaining fixture:\n");
  console.table(
    cfg.precomputedFixtures.map(([homeIdx, awayIdx, pHome, pDraw, pAway]) => ({
      Match: `${cfg.teams[homeIdx]} vs ${cfg.teams[awayIdx]} `,
      "Home Win": pct(pHome),
      Draw: pct(pDraw),
      "Away Win": pct(pAway),
    }))
  );
}

(async function main() {
  console.log("Starting Monte Carlo simulation...");
  console.time("Simulation time");
  const results = await simulate();
  console.timeEnd("Simulation time");
  showTeamOdds(results);
  showMatchOdds();
})();

File: 2026-qualifier/simulate.go
// simulate.go
// Monte‑Carlo World‑Cup‑2026 qualifier simulator (UEFA Group H)
// Uses a single JSON file (groupH.json) for **all** inputs:
//   • teams, Elo ratings, current table
//   • remaining fixtures
//   • model parameters (draw‑R, home bonus, playoff win prob)
//   • numberOfSimulations  ← NEW
//
// Build / run:
//   go run simulate.go
// ---------------------------------------------------------------------------

package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"math"
	"math/rand"
	"os"
	"path/filepath"
	"runtime"
	"sync"
	"text/tabwriter"
	"time"

	yaml "gopkg.in/yaml.v2"
)

/* -------------------------------------------------------------------------
   Config loader
-------------------------------------------------------------------------- */

var prngChoice = flag.String("prng", "xorshift32", "PRNG to use: 'math' or 'xorshift32'")

type Config struct {
	NumberOfSimulations int                `json:"numberOfSimulations" yaml:"numberOfSimulations"`
	Teams               []string           `json:"teams" yaml:"teams"`
	Elo                 map[string]float64 `json:"elo" yaml:"elo"`
	HomeBonus           float64            `json:"homeBonus" yaml:"homeBonus"`
	CurrentPoints       map[string]int     `json:"currentPoints" yaml:"currentPoints"`
	Fixtures            [][2]string        `json:"fixtures" yaml:"fixtures"`
	DrawR               float64            `json:"drawR" yaml:"drawR"`
	PlayoffWinProb      float64            `json:"playoffWinProb" yaml:"playoffWinProb"`

	// Precomputed data
	TeamIdx             map[string]int
	BasePts             []int
	PrecomputedFixtures [][5]float64 // home_idx, away_idx, p_home, p_draw, p_away
}

func loadConfig() Config {
	flag.Parse()
	cfgPath := flag.Arg(0)
	if cfgPath == "" {
		cfgPath = "groupH.json"
	}

	raw, err := os.ReadFile(cfgPath)
	if err != nil {
		log.Fatalf("can’t read %s: %v", cfgPath, err)
	}
	var c Config

	ext := filepath.Ext(cfgPath)
	switch ext {
	case ".json":
		if err := json.Unmarshal(raw, &c); err != nil {
			log.Fatalf("bad JSON: %v", err)
		}
	case ".yaml", ".yml":
		if err := yaml.Unmarshal(raw, &c); err != nil {
			log.Fatalf("bad YAML: %v", err)
		}
	default:
		log.Fatalf("unsupported config file format: %s", ext)
	}

	if c.NumberOfSimulations <= 0 {
		c.NumberOfSimulations = 1_000_000
	}

	if len(c.Teams) == 0 {
		log.Fatalf("config error: 'teams' cannot be empty")
	}
	if len(c.Elo) == 0 {
		log.Fatalf("config error: 'elo' cannot be empty")
	}
	if c.HomeBonus == 0 {
		log.Fatalf("config error: 'homeBonus' cannot be zero")
	}
	if len(c.CurrentPoints) == 0 {
		log.Fatalf("config error: 'currentPoints' cannot be empty")
	}
	if len(c.Fixtures) == 0 {
		log.Fatalf("config error: 'fixtures' cannot be empty")
	}
	if c.DrawR == 0 {
		log.Fatalf("config error: 'drawR' cannot be zero")
	}
	if c.PlayoffWinProb == 0 {
		log.Fatalf("config error: 'playoffWinProb' cannot be zero")
	}

	// Precompute team indices and base points
	c.TeamIdx = make(map[string]int, len(c.Teams))
	c.BasePts = make([]int, len(c.Teams))
	for i, team := range c.Teams {
		c.TeamIdx[team] = i
		c.BasePts[i] = c.CurrentPoints[team]
	}

	// Precompute fixture odds with integer indices
	c.PrecomputedFixtures = make([][5]float64, len(c.Fixtures))
	for i, f := range c.Fixtures {
		home, away := f[0], f[1]
		homeIdx, awayIdx := c.TeamIdx[home], c.TeamIdx[away]

		delta := c.Elo[home] + c.HomeBonus - c.Elo[away]
		pDraw := c.drawProb(delta)
		pHome := (1 - pDraw) * eloWin(c.Elo[home]+c.HomeBonus, c.Elo[away])
		pAway := 1 - pHome - pDraw

		c.PrecomputedFixtures[i] = [5]float64{float64(homeIdx), float64(awayIdx), pHome, pDraw, pAway}
	}

	return c
}

var cfg = loadConfig()

/* -------------------------------------------------------------------------
   Probability helpers
-------------------------------------------------------------------------- */

func eloWin(a, b float64) float64 { return 1 / (1 + math.Pow(10, (b-a)/400)) }

func (c *Config) drawProb(delta float64) float64 {
	w := 1 / (1 + math.Pow(10, -delta/400))
	return 2 * w * (1 - w) * c.DrawR
}

/* -------------------------------------------------------------------------
   Simulation
-------------------------------------------------------------------------- */

type tallies struct{ direct, playoff, fail int64 }

// PRNG interface
type prng interface {
	nextFloat64() float64
}

// xorshift32 PRNG
type xorshift32 struct {
	state uint32
}

func newXorshift32(seed int64) *xorshift32 {
	return &xorshift32{state: uint32(seed)}
}

func (r *xorshift32) nextFloat64() float64 {
	r.state ^= r.state << 13
	r.state ^= r.state >> 17
	r.state ^= r.state << 5
	return float64(r.state) / 4294967296.0
}

// Wrapper for math/rand to satisfy the prng interface
type mathRand struct {
	*rand.Rand
}

func (r *mathRand) nextFloat64() float64 {
	return r.Float64()
}

func simulate() []tallies {
	numWorkers := runtime.NumCPU()
	simsPerWorker := cfg.NumberOfSimulations / numWorkers
	remainingSims := cfg.NumberOfSimulations % numWorkers
	resultsChan := make(chan []tallies, numWorkers)
	var wg sync.WaitGroup

	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			localCounts := make([]tallies, len(cfg.Teams))

			var r prng
			seed := time.Now().UnixNano() + int64(workerID)
			if *prngChoice == "math" {
				r = &mathRand{rand.New(rand.NewSource(seed))}
			} else {
				r = newXorshift32(seed)
			}

			pts := make([]int, len(cfg.Teams))

			numSims := simsPerWorker
			if workerID < remainingSims {
				numSims++
			}

			for s := 0; s < numSims; s++ {
				copy(pts, cfg.BasePts)

				for _, f := range cfg.PrecomputedFixtures {
					homeIdx, awayIdx := int(f[0]), int(f[1])
					pHome, pDraw := f[2], f[3]
					randVal := r.nextFloat64()
					if randVal < pHome {
						pts[homeIdx] += 3
					} else if randVal < pHome+pDraw {
						pts[homeIdx]++
						pts[awayIdx]++
					} else {
						pts[awayIdx] += 3
					}
				}

				first, second := 0, 1
				if pts[1] > pts[0] || (pts[1] == pts[0] && r.nextFloat64() < 0.5) {
					first, second = 1, 0
				}
				for i := 2; i < len(cfg.Teams); i++ {
					if pts[i] > pts[first] || (pts[i] == pts[first] && r.nextFloat64() < 0.5) {
						second = first
						first = i
					} else if pts[i] > pts[second] || (pts[i] == pts[second] && r.nextFloat64() < 0.5) {
						second = i
					}
				}

				localCounts[first].direct++
				localCounts[second].playoff++
				for i := 0; i < len(cfg.Teams); i++ {
					if i != first && i != second {
						localCounts[i].fail++
					}
				}
			}
			resultsChan <- localCounts
		}(i)
	}

	wg.Wait()
	close(resultsChan)

	finalCounts := make([]tallies, len(cfg.Teams))
	for workerResult := range resultsChan {
		for i, counts := range workerResult {
			finalCounts[i].direct += counts.direct
			finalCounts[i].playoff += counts.playoff
			finalCounts[i].fail += counts.fail
		}
	}

	return finalCounts
}

/* -------------------------------------------------------------------------
   Pretty print helpers
-------------------------------------------------------------------------- */

func pct(x float64) string { return fmt.Sprintf("%.1f%%", x*100) }

/* -------------------------------------------------------------------------
   Main
-------------------------------------------------------------------------- */

func main() {
	start := time.Now()

	count := simulate()
	elapsed := time.Since(start)
	fmt.Printf("Simulation time: %v (%d simulations)\n\n", elapsed, cfg.NumberOfSimulations)

	// Table 1 – qualification probabilities
	w := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
	fmt.Fprintln(w, "Team\tDirect\tPlayoff\tEliminated\tOverall")
	for i, t := range cfg.Teams {
		c := count[i]
		d := float64(c.direct) / float64(cfg.NumberOfSimulations)
		p := float64(c.playoff) / float64(cfg.NumberOfSimulations)
		f := float64(c.fail) / float64(cfg.NumberOfSimulations)
		overall := d + p*cfg.PlayoffWinProb
		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n",
			t, pct(d), pct(p), pct(f), pct(overall))
	}
	w.Flush()
	fmt.Println()

	// Table 2 – per‑fixture odds
	w2 := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', 0)
	fmt.Fprintln(w2, "Match\tHome Win\tDraw\tAway Win")
	for _, f := range cfg.PrecomputedFixtures {
		homeIdx, awayIdx := int(f[0]), int(f[1])
		pHome, pDraw, pAway := f[2], f[3], f[4]
		fmt.Fprintf(w2, "%s vs %s\t%s\t%s\t%s\n",
			cfg.Teams[homeIdx], cfg.Teams[awayIdx], pct(pHome), pct(pDraw), pct(pAway))
	}
	w2.Flush()
}
